### PR TITLE
Support mutable Collection and Maps in the default PropertyAccessor

### PR DESCRIPTION
**Pull request by rozza**
If a property is a Collection or a Map and has no setter, try using
the getter to mutate the Collection or Map.

JAVA-2648

### CODE REVIEW

File: bson/src/test/unit/org/bson/codecs/pojo/PojoCustomTest.java [hunk 1]
```diff
@@ -387,4 +389,16 @@ public void testInvalidGetterAndSetterModelEncoding() {
     public void testInvalidGetterAndSetterModelDecoding() {
         decodingShouldFail(getCodec(InvalidGetterAndSetterModel.class), "{'integerField': 42, 'stringField': 'myString'}");
     }
+
+    @Test(expected = CodecConfigurationException.class)
+    public void testImmutableFinalMapAndCollectionsModel() {
```
Comment (rozza) [inline bson/src/test/unit/org/bson/codecs/pojo/PojoCustomTest.java]: This proves the asymmetry when dealing with immutable collections / maps and shows that the propertyAccessor doesn't leak an error when trying to mutate.

**Comment by rozza** [comment]
I'm unsure this is a good idea, it feels hackish to me and only works for mutable collections / maps.
If the user has a private collection / map then why should we try to set it.

The reason Jackson works is because it sets the field directly even if its private - which is something we dont do.

**Comment by jyemin** [comment]
In the models that work with Jackson, the expectation would be that in normal application use one would write:

     Person p = new Person();
     p.setName("Jeff");
     p.getPets().add(new Pet("dog"));
     p.getPets().add(new Pet("cat"));

?  If so, this PR seems reasonable even if the implementation in Jackson is a bit different than ours.

**Comment by rozza** [comment]
Jackson does not use the getter - rather sets the field directly

**Comment by rozza** [comment]
Closing as we opted not to mutate private fields and using the getters to mutate them seems wrong.

**Pull request closed without merging**